Kickstarter-iOS/Views/Controllers/RewardsCollectionViewController.swift

```diff
@@ -0,0 +1,127 @@
+import Foundation
+import Library
+import KsApi
+import Prelude
+
+final class RewardsCollectionViewController: UICollectionViewController {
+
+  // MARK: - Properties
+
+  private let dataSource = RewardsCollectionViewDataSource()
+  private let viewModel = RewardsCollectionViewModel()
+
+  private let hiddenPagingScrollView: UIScrollView = {
+    UIScrollView()
+      |> \.isPagingEnabled .~ true
+      |> \.isHidden .~ true
+  }()
+
+  private let layout: UICollectionViewFlowLayout = {
+    UICollectionViewFlowLayout()
+      |> \.minimumLineSpacing .~ Styles.grid(3)
+      |> \.sectionInset .~ .init(all: Styles.grid(6))
+      |> \.scrollDirection .~ .horizontal
+  }()
+
+  private var flowLayout: UICollectionViewFlowLayout? {
+    return self.collectionView.collectionViewLayout as? UICollectionViewFlowLayout
+  }
+
+  private let peekAmountInset = Styles.grid(3)
+
+  static func instantiate(with project: Project, refTag: RefTag?) -> RewardsCollectionViewController {
+    let rewardsCollectionVC = RewardsCollectionViewController()
+    rewardsCollectionVC.viewModel.inputs.configure(with: project, refTag: refTag)
+
+    return rewardsCollectionVC
+  }
+
+  init() {
+    super.init(collectionViewLayout: self.layout)
+
+    let closeButton = UIBarButtonItem(image: UIImage(named: "icon--cross"),
```

Comment:
I've seen this pattern in our code base (`VC` performing presentation/dismissal and other workflow actions) and was wondering if this is something we could start chatting around? Cc: @justinswart @cdolm92 @Scollaco 

This approach has been mostly used by Apple in their demo projects (mostly for simplicity of the demo) but it gives the controller powers (knowledge) it shouldn't really know about. In other words the controller should not know how to dismiss itself (it should be the role of it's presenting/parent view controller or ideally something like flow controller). It's a long way to an ideal state (flow controllers), or other patterns, but we could at least start talking whether we'd want to inject these actions from outside of the controller so that it's implementation agnostic (this is a common practice when defining React Native components for example and passing their actions from the parent component).

Example would be

```swift
// in parent view controller
let vc = RewardsCollectionViewController.instantiate()
vc.dismissHandler = { [weak self] in
  self?.dismiss(animated: true)
}

let nc = UINavigationController(rootViewController: vc)
self.present(nc, animated: true)
```

or

```swift
let vc = RewardsCollectionViewController.instantiate()
vc.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "icon--cross", style: .plain, target: self, action: #selector(ParentViewController.closeButtonTapped))

let nc = UINavigationController(rootViewController: vc)
self.present(nc, animated: true)
```

This way this controller can either be presented modally or pushed on the navigation stack and by providing dismissal action it could either be dismissed or popped. If we didn't inject the action based on the workflow / presentation context imaging we pushed this on the stack...the internally declared dismissal would not work because it only assumes modal presentation (this removes reusability and flexibility).

Reply:
I think this is something we should discuss as a team & approach more holistically in our codebase if it's something we want to move towards. I think it's a good topic to bring up for sure though, and will become more relevant as our app gets larger.

Reply:
Yeah, worth exploring further but outside the scope of this PR. While it's true that we may want to have dismissal stuff be declared at the presentation callsite, we should just be smart about how we do that. For example, I would want to avoid a situation where a child is calling to a parent via delegate methods to have something presented. UIKit happily presents VC's from any VC in the stack and I don't necessarily think that's a problem. What I'm saying is that the delegate pattern is bad for this but potentially exploring a better abstraction like a Presenter or Flow controller may be worthwhile.